js/src/wrappers/claude-agent-sdk.ts

```diff
@@ -0,0 +1,505 @@
+} from "./anthropic-tokens-util";
+import { getNumberProperty } from "./ai-sdk-shared";
+
+/**
+ * Types from @anthropic-ai/claude-agent-sdk
+ */
+type SDKMessage = {
+  type: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [key: string]: any;
+};
+
+type QueryOptions = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [key: string]: any;
+};
+
+type Query = AsyncGenerator<SDKMessage, void, unknown> & {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [key: string]: any;
+};
+
+type CallToolResult = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  content: Array<any>;
+  isError?: boolean;
+};
+
+type ToolHandler<T> = (
+  args: T,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  extra: any,
+) => Promise<CallToolResult>;
+
+type SdkMcpToolDefinition<T> = {
+  name: string;
+  description: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  inputSchema: any;
+  handler: ToolHandler<T>;
+};
+
+/**
+ * Wraps the Claude Agent SDK's query function to add Braintrust tracing.
+ * Traces the entire agent interaction including all streaming messages.
+ *
+ * @param queryFn - The original query function from @anthropic-ai/claude-agent-sdk
+ * @returns A wrapped query function with tracing
+ *
+ * @example
+ * ```typescript
+ * import { query } from "@anthropic-ai/claude-agent-sdk";
+ * import { wrapClaudeAgentQuery } from "braintrust";
+ *
+ * const wrappedQuery = wrapClaudeAgentQuery(query);
+ *
+ * const result = wrappedQuery({
+ *   prompt: "Write a hello world function",
+ *   options: { model: "claude-3-5-sonnet-20241022" }
+ * });
+ *
+ * for await (const message of result) {
+ *   console.log(message);
+ * }
+ * ```
+ */
+export function wrapClaudeAgentQuery<
+  T extends (...args: unknown[]) => AsyncGenerator<SDKMessage, void, unknown>,
+>(queryFn: T, defaultThis?: unknown): T {
+  const proxy: T = new Proxy(queryFn, {
+    apply(target, thisArg, argArray) {
+      const params = (argArray[0] ?? {}) as {
+        prompt?: string | AsyncIterable<SDKMessage>;
+        options?: QueryOptions;
+      };
+
+      const { prompt, options = {} } = params;
+
+      const span = startSpan({
+        name: "Claude Agent",
+        spanAttributes: {
+          type: SpanTypeAttribute.TASK,
+        },
+        event: {
+          input:
+            typeof prompt === "string"
+              ? prompt
+              : { type: "streaming", description: "AsyncIterable<SDKMessage>" },
+          metadata: options,
+        },
+      });
+
+      let ended = false;
+      const finalResults: Array<{ content: unknown; role: string }> = [];
+
+      // Track messages by message.message.id to group streaming updates
+      let currentMessageId: string | undefined;
+      let currentMessageStartTime = getCurrentUnixTimestamp();
+      const currentMessages: SDKMessage[] = [];
+
+      const endSpan = () => {
```

Comment:
#nit I seems silly to have a separate method. I also wonder what span.end(); span.end() will do. Is it not idempotent?